### PR TITLE
Implement usermap exponential backoff (INF-1310)

### PR DESCRIPTION
**Pull request by williamnswanson**
Update usermap script to use exponential backoff. The minimum and maximum timeouts can be set with options.

### CODE REVIEW

File: osg-comanage-project-usermap.py [hunk 1]
```diff
@@ -50,6 +55,8 @@ class Options:
     outfile = None
     authstr = None
     filtergrp = None
+    minTimeout = MINTIMEOUT
+    maxTimeout = MAXTIMEOUT
```
Comment (brianhlin) [inline osg-comanage-project-usermap.py]:
```suggestion
    min_timeout = MINTIMEOUT
    max_timeout = MAXTIMEOUT
```

Attempting to align with PEP8: https://peps.python.org/pep-0008/#function-and-variable-names

File: osg-comanage-project-usermap.py [hunk 2]
```diff
@@ -159,13 +179,15 @@ def parse_options(args):
...
-        if op == '-g': options.filtergrp = arg
+        if op == '-u': options.user       = arg
+        if op == '-c': options.osg_co_id  = int(arg)
+        if op == '-d': passfd             = int(arg)
+        if op == '-f': passfile           = arg
+        if op == '-e': options.endpoint   = arg
+        if op == '-o': options.outfile    = arg
+        if op == '-g': options.filtergrp  = arg
+        if op == '-t': options.minTimeout = float(arg)
+        if op == '-T': options.maxTimeout = float(arg)
```
Comment (brianhlin) [inline osg-comanage-project-usermap.py]:
```suggestion
        if op == '-t': options.min_timeout = float(arg)
        if op == '-T': options.max_timeout = float(arg)
```

**Review by brianhlin — DISMISSED** [review]
This would get the job done but I'm more interested in actually implementing an exponential timeout: try 5, 25, 125, and 625 seconds and then give up. I'm also fine if it's not configurable via option

File: osg-comanage-project-usermap.py [hunk 3]
```diff
@@ -87,8 +94,21 @@ def mkrequest(target, **kw):
...
+    while trying:
+        try:
+            resp = urllib.request.urlopen(req, timeout=currentTimeout)
+            payload = resp.read()
+            trying = False
+        except urllib.error.URLError as exception:
+            if (currentTimeout < options.maxTimeout):
+                currentTimeout *= TIMEOUTMULTIPLE
+            else:
+                print("Exception raised after maximum timeout " + str(options.maxTimeout) + " seconds reached. Exception reason: " + str(exception.reason) + ".\n Request: " + str(req.full_url))
```
Comment (brianhlin) [inline osg-comanage-project-usermap.py]: You could just use f-strings here

File: osg-comanage-project-usermap.py [hunk 4]
```diff
@@ -87,8 +94,21 @@ def mkrequest(target, **kw):
...
+                trying = False
+                quit()
```
Comment (brianhlin) [inline osg-comanage-project-usermap.py]: You should use `sys.exit()` instead. Also, no need to set `trying` if you raise the SystemExit exception via your exit function. Also you can simplify all this to one line by adding your error message as the arg to `sys.exit()`.

**Review by brianhlin — COMMENTED** [review]
Oh wait, I guess that is exponential! It's a little complicated but it'll do

**Comment by brianhlin** [comment]
I also updated the PR title (see https://osg-htc.org/technology/software/git-software-development/#summarize-your-commits-in-the-pull-request-title), partially because the title makes its way into the merge commit so we want it to summarize the changes in a human-readable way. Also, Jira will look for the ticket key in commits/title so you want that to be well-formatted

File: osg-comanage-project-usermap.py (hunk 1 above)
Comment (williamnswanson) [reply osg-comanage-project-usermap.py]: I've incorporated this change into commit 80e513c

File: osg-comanage-project-usermap.py (hunk 2 above)
Comment (williamnswanson) [reply osg-comanage-project-usermap.py]: I've incorporated this change in 80e513c

File: osg-comanage-project-usermap.py (hunk 3 above)
Comment (williamnswanson) [reply osg-comanage-project-usermap.py]: Replaced the string addition with f-strings in 80e513c.

File: osg-comanage-project-usermap.py (hunk 4 above)
Comment (williamnswanson) [reply osg-comanage-project-usermap.py]: Replaced quit() with sys.exit() with the f-string passed in commit 80e513c

**Review by brianhlin — APPROVED** [review]
LGTM!

File: osg-comanage-project-usermap.py [hunk 5]
```diff
@@ -87,8 +94,20 @@ def mkrequest(target, **kw):
...
+        try:
+            resp = urllib.request.urlopen(req, timeout=currentTimeout)
+            payload = resp.read()
+            trying = False
+        except urllib.error.URLError as exception:
+            if (currentTimeout < options.max_timeout):
+                currentTimeout *= TIMEOUTMULTIPLE
+            else:
+                sys.exit(f"Exception raised after maximum timeout {options.max_timeout} seconds reached. "
+                         + "Exception reason: {exception.reason}.\n Request: {req.full_url}")
```
Comment (brianhlin) [inline osg-comanage-project-usermap.py]:
```suggestion
                         + f"Exception reason: {exception.reason}.\n Request: {req.full_url}")
```

Oh hey, flake8 caught something!

**Pull request merged**